### PR TITLE
review: feature: Unnamed Variables & Patterns (JEP 456)

### LINKED ISSUE #5722: java 22
For the record

Java 22 was just released along with JavaFX 22 [1][2]. Thank you to all the projects who contributed to those releases by testing and providing feedback using their respective early-access builds. And to celebrate that, the Java DevRel Team hosted a +4h live-stream with guests such as Brian Goetz, Viktor Klang, Alan Bateman, etc. You can watch the launch stream replay here [3].
 
The JDK 23 schedule is now known [4] with rampdown starting early June and general availability sets for mid-September. So far, 2 JEPs have been targeted to JDK 23:
- JEP 455: Primitive Types in Patterns, instanceof, and switch (Preview) [5]
- JEP 466: Class-File API (2nd Preview) [6]
 
The focus should now be shifted to testing your project(s) on JDK 23. And don't forget that the Oracle setup-java github action [7] supports, amongst others, the latest OpenJDK 23 Early-Access builds. So, JDK 23 EA testing is literally one pipeline away.
 
[1] https://mail.openjdk.org/pipermail/jdk-dev/2024-March/008827.html
[3] https://jdk.java.net/javafx22/
[3] https://www.youtube.com/live/AjjAZsnRXtE?feature=shared&t=278
[4] https://openjdk.org/projects/jdk/23/
[5] https://openjdk.org/jeps/455
[6] https://openjdk.org/jeps/466
[7] https://github.com/oracle-actions/setup-java
 
 
## Heads-up: JDK 20-23: Support for Unicode CLDR Version 42
 
The JDK update to CLDR version 42 included a change where regular spaces in date/time formats (and some other formatted values) were replaced with (narrow) non-breaking spaces. This lead to issues for existing code that relied on parsing such strings. To address that, JDK 23 allows loose matching of spaces when parsing date/time strings. Loose matching is performed in the lenient parsing style for both date/time parsers in `java.time.format` and `java.text` packages. In the default strict parsing style, those spaces are considered distinct as before.
 
Please read this updated heads-up [9] for details on how to configure strict/lenient parsing in the `java.time.format` (strict by default) and `java.text` (lenient by default) packages.
 
[9] https://inside.java/2024/03/29/quality-heads-up/
 
 
## Heads-up: macOS 14 users running on Apple silicon systems should update directly to macOS 14.4.1
 
An issue introduced by macOS 14.4 caused some Java processes, regardless of the Java version, to terminate unexpectedly on Apple silicon (AArch64). On March 25 Apple released macOS 14.4.1 and indicated on their support site that it addresses this issue. Oracle can confirm that after applying macOS 14.4.1 we are unable to reproduce the problem. So, Java users on macOS 14 running on Apple silicon systems should skip macOS 14.4 and update directly to macOS 14.4.1.
 
More details can be found on https://blogs.oracle.com/java/post/java-on-macos-14-4
 
 
## JDK 23 Early-Access Builds
 
The JDK 23 EA builds 16 are available [10], and are provided under the GNU General Public License v2, with the Classpath Exception. The Release Notes [11] are also available.
 
### Changes in recent JDK 23 builds that may be of interest:
- JDK-8324774: Add DejaVu web fonts (reported by AssertJ)
- JDK-8327385: Add JavaDoc option to exclude web fonts from generated documentation (reported by AssertJ)
- JDK-8328638: Fallback option for POST-only OCSP requests
- JDK-8320362: Load anchor certificates from Keychain keystore
- JDK-8327875: ChoiceFormat should advise throwing UnsupportedOperationException for unused methods
- JDK-8296244: Alternate implementation of user-based authorization Subject APIs that doesn’t depend on Security Manager APIs
- JDK-8327818: Implement Kerberos debug with sun.security.util.Debug
- JDK-7036144: GZIPInputStream readTrailer uses faulty available() test for end-of-stream
- JDK-8319251: Change LockingMode default from LM_LEGACY to LM_LIGHTWEIGHT
- JDK-8327651: Rename DictionaryEntry members related to protection domain
- JDK-8321408: Add Certainly roots R1 and E1
- JDK-8164094: javadoc allows to create a @link to a non-existent method
- JDK-8325496: Make TrimNativeHeapInterval a product switch
- JDK-8174269: Remove COMPAT locale data provider from JDK
- JDK-8322750: Test "api/java_awt/interactive/SystemTrayTests.html" failed because …
- JDK-8139457: Relax alignment of array elements
- JDK-8256314: JVM TI GetCurrentContendedMonitor is implemented incorrectly
- JDK-8326908: DecimalFormat::toPattern throws OutOfMemoryError when pattern is empty string
- JDK-8247972: incorrect implementation of JVM TI GetObjectMonitorUsage
- JDK-8325580: Remove "alternatives --remove" call from Java rpm installer
- JDK-8326838: JFR: Native mirror events
- JDK-8326106: Write and clear stack trace table outside of safepoint
- JDK-8323183: ClassFile API performance improvements
- JDK-8324829: Uniform use of synchronizations in NMT
- JDK-8326586: Improve Speed of System.map
- JDK-8318761: MessageFormat pattern support for CompactNumberFormat, ListFormat, and DateTimeFormatter
- JDK-8265372: Simplify PKCS9Attribute
- JDK-8317431: Implement simpler Comparator when building certification paths
- JDK-8327093: Add truncate function to BitMap API
 
Note: A more detailed list of changes can be found here [12].
 
[10] https://jdk.java.net/23/
[11] https://jdk.java.net/23/release-notes
[12] https://github.com/openjdk/jdk/compare/jdk-23+10...jdk-23+16
 
 
## JavaFX 23 Early-Access Builds
 
These are early access builds of the JavaFX 23 Runtime built from openjdk/jfx [13]. These builds enable JavaFX application developers to build and test their applications with JavaFX 23 on JDK 23. Although these builds are designed to work with JDK 23, they are also known to work with JDK 17 and later versions.
 
The latest early access builds of JavaFX 23 Builds 10 (2024/3/26) are available [14], under the GNU General Public License, version 2, with the Classpath Exception.
 
[13] https://github.com/openjdk/jfx
[14] https://jdk.java.net/javafx23/

### PR DESCRIPTION
**Pull request by SirYwell**
Implementation of [JEP 456](https://openjdk.org/jeps/456).

It mainly consists of two features:

1. Unnamed patterns (`_`)
2. Unnamed (local) variables (`Type _`)

Fixes #5722.

### CODE REVIEW

File: src/main/java/spoon/reflect/declaration/CtParameter.java [hunk 1]
```diff
@@ -60,6 +60,12 @@ public interface CtParameter<T> extends CtVariable<T>, CtShadowable {
 	@UnsettableProperty
 	<C extends CtVariable<T>> C setDefaultExpression(CtExpression<T> assignedExpression);
 
+	/**
+	 * {@return whether this parameter is <a href="https://openjdk.org/jeps/456">unnamed</a>}
+	 */
+	@DerivedProperty
+	boolean isUnnamed();
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/reflect/declaration/CtParameter.java:67]: Method parameters can not be unnamed. Is this for Lambdas? Maybe clarify that in the docs?

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 2]
```diff
@@ -2383,4 +2385,9 @@ public void visitCtReceiverParameter(CtReceiverParameter receiverParameter) {
 			printer.writeSeparator("this");
 		}
 	}
+
+	@Override
+	public void visitCtUnnamedPattern(CtUnnamedPattern unnamedPattern) {
+		printer.writeKeyword(CtLocalVariableImpl.UNNAMED_VARIABLE_NAME);
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java:2391]: Why does this constant live in CtLocalVariableImpl and not an interface? Multiple places can be unnamed (e.g. also catch variables which somehow are not local variables)...

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java [hunk 3]
```diff
@@ -1358,9 +1358,11 @@ public boolean visit(LabeledStatement labeledStatement, BlockScope scope) {
 
 	@Override
 	public boolean visit(LocalDeclaration localDeclaration, BlockScope scope) {
+		// an unnamed pattern does not have a type in the source, but the binding holds the inferred type
+		CtElement element;
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java:1362]: Why is this declared up here if it is only used and assigned once, much further down?

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java [hunk 4]
```diff
@@ -1754,9 +1757,22 @@ public boolean visit(GuardedPattern guardedPattern, BlockScope scope) {
...
+			pattern = factory.Core().createTypePattern();
+		}
+		context.enter(pattern, anyPattern);
+		// when we have an unnamed pattern, we don't want to visit the local variable declaration anymore
+		return !unnamedPattern;
+	}
+
+	// only '_', NOT 'Type _'
+	private boolean isUnnamedPattern(TypePattern pattern) {
+		return pattern.isUnnamed() && pattern.local.type == null && pattern.local.binding != null;
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java:1775]: Cool, so JDT differs from the spec here? :^)

File: src/main/java/spoon/support/compiler/jdt/ParentExiter.java [hunk 5]
```diff
@@ -517,6 +517,23 @@ public <E> void visitCtCase(CtCase<E> caseStatement) {
...
+		if (cs.constantExpressions == null) {
+			return false;
+		}
+		if (child instanceof CtExpression
+			&& caseStatement.getCaseExpressions().size() < cs.constantExpressions.length) {
+			return true;
+		}
+		if (cs.constantExpressions.length == 1 && cs.constantExpressions[0] instanceof EitherOrMultiPattern eomp) {
+			return caseStatement.getCaseExpressions().size() < eomp.getAlternatives().length;
+		}
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/support/compiler/jdt/ParentExiter.java:533]: What does this do? Maybe add a sentence or two to explain what this does?

File: src/main/java/spoon/support/reflect/code/CtCatchVariableImpl.java [hunk 6]
```diff
@@ -107,6 +107,11 @@ public <C extends CtVariable<T>> C setDefaultExpression(CtExpression<T> defaultE
 		return (C) this;
 	}
 
+	@Override
+	public boolean isUnnamed() {
+		return CtLocalVariableImpl.isUnnamed(this);
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/support/reflect/code/CtCatchVariableImpl.java:112]: Same here, why is this method on the impl and not the interface?

File: src/main/java/spoon/support/reflect/declaration/CtParameterImpl.java [hunk 7]
```diff
@@ -78,6 +79,10 @@ public <C extends CtVariable<T>> C setDefaultExpression(CtExpression<T> defaultE
 		return (C) this;
 	}
 
+	@Override
+	public boolean isUnnamed() {
+		return CtLocalVariableImpl.isUnnamed(this);
+	}
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/support/reflect/declaration/CtParameterImpl.java:85]:
```suggestion
	}

```

File: src/test/java/spoon/test/pattern/RecordPatternTest.java [hunk 8]
```diff
@@ -166,6 +169,23 @@ void testTypePatternWithVar(String var) {
 		assertEquals(pattern.contains("final"), variable.isFinal());
 	}
 
+	@Test
+	void testUnnamedPatternInRecordPattern() {
+		CtSwitch<?> ctSwitch = createFromSwitch("var i = 1; record Int(int i) {}", "Int(_)");
```
Comment (I-Al-Istannen) [inline src/test/java/spoon/test/pattern/RecordPatternTest.java:174]: Maybe add a contract here?

File: src/test/java/spoon/test/variable/VariableTest.java [hunk 9]
```diff
@@ -191,4 +200,27 @@ public void testVarInLambda() {
         assertEquals("java.lang.Long", lambda.getParameters().get(1).getType().getQualifiedName());
         assertEquals("(var x,var y) -> x + y", lambda.toString()); // we should print var, if it was in the original code
     }
+
+    @ModelTest(value = "./src/test/resources/spoon/test/unnamed/UnnamedVar.java", complianceLevel = 22)
+    void testUnnamedVariable(Factory factory) throws IOException {
```
Comment (I-Al-Istannen) [inline src/test/java/spoon/test/variable/VariableTest.java:205]: Contract? I have no idea what this test tests :P That the lines match the source and everything is unnamed?

File: src/test/java/spoon/test/variable/VariableTest.java [hunk 10]
```diff
@@ -191,4 +200,27 @@ public void testVarInLambda() {
...
+			} else if (variable instanceof CtParameter<?> v) {
+				assertTrue(v.isUnnamed());
+			} else if (variable instanceof CtCatchVariable<?> v) {
+				assertTrue(v.isUnnamed());
+			}
+            assertThat(variable).getPosition().isNotEqualTo(SourcePosition.NOPOSITION);
+            String line = lines.get(variable.getPosition().getLine() - 1);
+            assertThat(line).contains(variable.toString());
+        }
+    }
```
Comment (I-Al-Istannen) [inline src/test/java/spoon/test/variable/VariableTest.java:225]: Something in the indentation here is messed up and mixes spaces and tabs.

File: src/test/java/spoon/test/variable/VariableTest.java [hunk 11]
```diff
@@ -191,4 +200,27 @@ public void testVarInLambda() {
         assertEquals("java.lang.Long", lambda.getParameters().get(1).getType().getQualifiedName());
         assertEquals("(var x,var y) -> x + y", lambda.toString()); // we should print var, if it was in the original code
     }
+
+    @ModelTest(value = "./src/test/resources/spoon/test/unnamed/UnnamedVar.java", complianceLevel = 22)
+    void testUnnamedVariable(Factory factory) throws IOException {
+        CtType<?> type = factory.Type().get("spoon.test.unnamed.UnnamedVar");
+        assertThat(type.getMethods()).isNotEmpty();
+        List<String> lines = java.nio.file.Files.readAllLines(type.getPosition().getFile().toPath());
```
Comment (I-Al-Istannen) [inline src/test/java/spoon/test/variable/VariableTest.java:208]: Do you use the FQN here by accident or necessity?

**Review by I-Al-Istannen — CHANGES_REQUESTED** [review]

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java (hunk 4 above)
Comment (SirYwell) [reply src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java:1775]: They're a bit lazy, they only have `TypePattern` with a variable that can be unnamed (`Type _` as a pattern), but represent the actual unnamed pattern by just not having a type set. `isUnnamed()` in their case refers to the variable, not the kind of pattern, from my understanding.

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java (hunk 2 above)
Comment (SirYwell) [reply src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java:2391]: I wasn't sure if we want to expose it directly. It's rather our internal representation, but maybe it shouldn't be used here then...

File: src/main/java/spoon/support/reflect/code/CtCatchVariableImpl.java (hunk 6 above)
Comment (SirYwell) [reply src/main/java/spoon/support/reflect/code/CtCatchVariableImpl.java:112]: The method holds code that works for all `CtVariable`s in the same way, but isn't really appropriate to be used on e.g. `CtField`. I thought hiding it in one of the implementations would make sense.

File: src/test/java/spoon/test/variable/VariableTest.java (hunk 11 above)
Comment (SirYwell) [reply src/test/java/spoon/test/variable/VariableTest.java:208]: `com.google.common.io.Files` is used in this class as well.

File: src/test/java/spoon/test/variable/VariableTest.java (hunk 10 above)
Comment (SirYwell) [reply src/test/java/spoon/test/variable/VariableTest.java:225]: Seems like this file is cursed.

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java (hunk 3 above)
Comment (SirYwell) [reply src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: It was a leftover from a different approach that failed miserably.

File: src/test/java/spoon/test/variable/VariableTest.java (hunk 10 above)
Comment (I-Al-Istannen) [reply src/test/java/spoon/test/variable/VariableTest.java]: Apparently it still uses spaces…

File: src/main/java/spoon/support/reflect/code/CtCatchVariableImpl.java (hunk 6 above)
Comment (I-Al-Istannen) [reply src/main/java/spoon/support/reflect/code/CtCatchVariableImpl.java:112]: Okay, I see. Maybe mark the method as `@Internal` then? People will use it if it is public.

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java (hunk 2 above)
Comment (I-Al-Istannen) [reply src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java:2391]: There is an `isUnnamed` getter, yea. The question is probably how people can create their own unnamed variables.

Comment (SirYwell) [reply src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: The constant actually lives in the interface, I just forgot that already when I used it here...

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 12]
```diff
@@ -124,6 +125,7 @@
 import spoon.reflect.reference.CtWildcardReference;
 import spoon.reflect.visitor.PrintingContext.Writable;
 import spoon.reflect.visitor.printer.CommentOffset;
+import spoon.support.reflect.code.CtLocalVariableImpl;
```
Comment (I-Al-Istannen) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java:128]:
```suggestion
```

**Review by I-Al-Istannen — CHANGES_REQUESTED** [review]

**Review by I-Al-Istannen — APPROVED** [review]

**Pull request merged**